content/docs/io/async_read_write.md

```diff
@@ -1,12 +1,311 @@
+            ReadExact::Reading { reader, buffer, .. } => Ok(Async::Ready((reader, buffer))),
+            ReadExact::Empty => panic!(),
+        }
+    }
+}
+```
+
+# Writing data with `AsyncWrite`
+
+Just like `poll_read` is the core piece of `AsyncRead`, `poll_write` is
+the core of `AsyncWrite`. Like `poll_read`, it maps the `WouldBlock`
+error that indicates that an I/O *`write`* operation would have blocked into
+`NotReady`, which again lets us interoperate with the world of
+futures. `AsyncWrite` also has a `poll_flush`, which provides an
+asynchronous analogue to [`Write`]'s `flush` method. The role of
+`poll_flush` is to make sure that any bytes previously written by
+`poll_write` are, well, flushed onto the underlying I/O resource
+(written out in network packets for example). Similar to `poll_write`, it
+wraps around `Write::flush`, and maps a `WouldBlock` error into
+`NotReady` to indicate that the flushing is still ongoing.
+
+`AsyncWrite`'s `poll_write` and `poll_flush` follow the same contract as
+`Future::poll` and `AsyncRead::poll_read`, namely that if they return
+`NotReady`, they have arranged for the current task to be notified when
+they can make progress again. Like with `poll_read`, this means that we
+can safely call these methods in our own futures, and know that we are
+also following the contract.
+
+Tokio uses the same mechanism to manage notifications for `poll_write`
+and `poll_flush` as it does for `poll_read`, and you can read more about
+it in the [non-blocking I/O] section of Tokio internals.
+
+## Shutdown
+
+`AsyncWrite` also adds one method that is *not* part of `Write`:
+`shutdown`. From [its documentation][shutdown]:
+
+> Initiates or attempts to shut down this writer, returning success when
+> the I/O connection has completely shut down.
+>
+> This method is intended to be used for asynchronous shutdown of I/O
+> connections. For example this is suitable for implementing shutdown of
+> a TLS connection or calling `TcpStream::shutdown` on a proxied
+> connection. Protocols sometimes need to flush out final pieces of data
+> or otherwise perform a graceful shutdown handshake, reading/writing
+> more data as appropriate. This method is the hook for such protocols
+> to implement the graceful shutdown logic.
+
+This sums `shutdown` up pretty nicely: it is a way to tell the writer
+that no more data is coming, and that it should indicate in whatever way
+the underlying I/O protocol requires. For TCP connections, for example,
+this usually entails closing the writing side of the TCP channel so that
+the other end receives and end-of-file in the form of a read that
+returns 0 bytes. You can often think of `shutdown` as what you _would_
+have done synchronously in the implementation of `Drop`; it's just that
+in the asynchronous world, you can't easily do something in `Drop`
+because you need to have an executor that keeps polling your writer!
+
+<!--
+Socket is safely dropped after shutdown returns Ok(Ready).
```

Comment:
Should this be deleted?

Reply:
It won't actually be visible in the rendered article since it's an HTML comment. I left it in there because it might be something we want to cover in later drafts.

Reply:
You might want to mark it as "Something we might want to cover in the future" - I'm not convinced future authors will know how to interpret the comment.

Reply:
Ah, yes, good point!